sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,393 @@
+
+    protected val attemptStack = new mutable.ArrayStack[K]()
+
+    override def hasNext: Boolean = input.hasNext || attemptStack.nonEmpty
+
+    protected def invokeFn(k: K): T = {
+      fn(k)
+    }
+
+    override def next(): T = {
+      // this is set on the first exception, and we add suppressed if there are others
+      // during the retry attempts
+      var lastException: Throwable = null
+      var firstAttempt: Boolean = true
+      var result: Option[T] = None
+      var doSplit = false
+      while ((input.hasNext || attemptStack.nonEmpty) &&
+          result.isEmpty) {
+        if (attemptStack.isEmpty && input.hasNext) {
+          attemptStack.push(input.next())
+        }
+        if (!firstAttempt) {
+          // call thread block API
+          RmmSpark.blockThreadUntilReady()
+        }
+        firstAttempt = false
+        val popped = attemptStack.pop()
+        val attempt =
+          if (doSplit) {
+            // If `split` OOMs, we are already the last thread standing
+            // there is likely not much we can do, and for now we don't handle
+            // this OOM
+            val splitted = splitAndClose(popped)
+            // the splitted sequence needs to be inserted in reverse order
+            // so we try the first item first.
+            splitted.reverse.foreach(attemptStack.push)
+            attemptStack.pop()
+          } else {
+            popped
+          }
+        doSplit = false
+        try {
+          // call the user's function
+          result = Some(invokeFn(attempt))
+        } catch {
+          case retryOOM: RetryOOM =>
+            if (lastException == null) {
+              lastException = retryOOM
+            } else {
+              retryOOM.addSuppressed(lastException)
+              lastException = retryOOM
+            }
+
+            // put it back
+            attemptStack.push(attempt)
+          case splitAndRetryOOM: SplitAndRetryOOM => // we are the only thread
+            if (lastException == null) {
+              lastException = splitAndRetryOOM
+            } else {
+              splitAndRetryOOM.addSuppressed(lastException)
+              lastException = splitAndRetryOOM
+            }
+            // put it back
+            attemptStack.push(attempt)
+            doSplit = true
+          case other: Throwable =>
+            if (lastException == null) {
+              lastException = other
+            } else {
+              other.addSuppressed(lastException)
+              lastException = other
+            }
+            // put this attempt back on our stack, so that it will be closed
+            attemptStack.push(attempt)
+
+            // we want to throw early here, since we got an exception
+            // we were not prepared to handle
+            throw lastException
+        }
+      }
+      result.get
+    }
+
+    // It is assumed that OOM in this function is not handled.
+    private def splitAndClose(item: K): Seq[K] = {
```

Comment:
I think this is fine for now, but we should file a follow on issue to explore what it would take to handle exceptions associated with string or list limits being violated in the output.  That is one case where we would want to do a split and there wouldn't be the single threaded guarantee that we get from the current code.

Reply:
https://github.com/NVIDIA/spark-rapids/issues/7816